### ISSUE TITLE
Mac

### ISSUE THREAD

**Issue by konsumer** [alifcommunity/webui#69]
I am trying to get it working on mac (x86_64 & arm64.) Am I missing a step?:

```
cd build/MacOS/Clang/
make
cd ../../..
clang build/MacOS/Clang/libwebui-2-aarch64-apple-darwin.a examples/C/minimal/main.c -Iinclude -o minimal
./minimal
```

This builds fine, but opens a blank page. I am having same trouble with the nim bindings.

Also, I notice it has `aarch64` (arm64) in name, even when I am on x86_64:

```
$ uname -a

Darwin konsumerbook.lan 22.3.0 Darwin Kernel Version 22.3.0: Mon Jan 30 20:42:11 PST 2023; root:xnu-8792.81.3~2/RELEASE_X86_64 x86_64

$ otool -hv build/MacOS/Clang/libwebui-2-aarch64-apple-darwin.a

Archive : build/MacOS/Clang/libwebui-2-aarch64-apple-darwin.a
build/MacOS/Clang/libwebui-2-aarch64-apple-darwin.a(webui.o):
Mach header
      magic  cputype cpusubtype  caps    filetype ncmds sizeofcmds      flags
MH_MAGIC_64   X86_64        ALL  0x00      OBJECT     5        776 SUBSECTIONS_VIA_SYMBOLS
build/MacOS/Clang/libwebui-2-aarch64-apple-darwin.a(mongoose.o):
Mach header
      magic  cputype cpusubtype  caps    filetype ncmds sizeofcmds      flags
MH_MAGIC_64   X86_64        ALL  0x00      OBJECT     5       1016 SUBSECTIONS_VIA_SYMBOLS
```

**Comment by hassandraga** [comment]
Thank you for reporting this. I updated the Makefile.

About the blank page, can you please share the log? I need to see the generated command that runs the browser. Additionally, if you build WebUI in debug mode, you can see the logs in the terminal and the dev-tools console in the browser.

I'm a Linux/Windows user. But I'm currently looking to buy a Mac just for the development and the testing of WebUI.

**Comment by hassandraga** [comment]
`make debug`

**Comment by konsumer** [comment]
`make debug` did not change output at all.

**Comment by konsumer** [comment]
An additional note: the default for Mac filesystem is case-insensitive, so `MacOS` and `macOS` both in `build/` will cause problems for most mac-people (I disable it when I set things up, because I hate that.) I would recommend putting the README in MacOS and deleting macOS.

**Comment by konsumer** [comment]
Running in `lldb` (debugger) I only get this output:

```
❯ lldb ./minimal
Current executable set to '/Users/konsumer/Desktop/otherdev/webui/minimal' (x86_64).
(lldb) r
Process 25465 launched: '/Users/konsumer/Desktop/otherdev/webui/minimal' (x86_64)
Process 25465 exited with status = 0 (0x00000000)
```

**Comment by hassandraga** [comment]
Thank you for the clarifications. Unfornantly I can't test it, I still need to get a Mac for now. And the MacOS Makefile is made by someone else. Usually, the _debug_ section should build the lib in debug mode + define `WEBUI_LOG`.

About the installation, yeah, you are right. I will soon add an install section for Linux and MacOS.

**Comment by konsumer** [comment]
I will play around with it, and see if I can get more info.

**Comment by konsumer** [comment]
Looking at the log, I ran this:

```
/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --user-data-dir="/Users/konsumer/.WebUI/WebUIChromeProfile" --no-first-run --disable-gpu --disable-software-rasterizer --no-proxy-server --safe-mode --disable-extensions --disable-background-mode --disable-plugins --disable-plugins-discovery --disable-translate --bwsi --disable-sync --incognito --app=https://google.com
```
and I get google, so I think the main problem uis `wait()` is not waiting and serving up `http://localhost:44136`

**Comment by hassandraga** [comment]
I don't see `[Thread] webui_server_start()` in your log. That's why `webui_wait()` fails. This is only related to MacOS implementation. I will look into it soon. Thank you for the report.

About the PR, if you think it fixes something or makes an improvement, yes, why not. I want you to know that contributions are welcome. I appreciate any help you can provide.

**Comment by konsumer** [comment]
> About the PR

Ok! The changes I made are in #70 

> I don't see [Thread] webui_server_start()

Ah, yes, that sounds right.

I also sent you an email about helping with more testing.

**Comment by hassandraga** [comment]
The `webui_wait()` for MacOS should be fixed now by https://github.com/alifcommunity/webui/commit/9a38611454ebbb1144a487af207e7a5aab0777ba

**Comment by konsumer** [comment]
It works! It pops up a window with custom content. It still holds the console when you exit, though.

**Comment by konsumer** [comment]
Actually, hmm, maybe I spoke too soon. It seems like it varies. sometimes it just returns, sometimes it runs but returns right away, and sometimes it holds the console.

**Comment by hassandraga** [comment]
Maybe it's another bug in MacOS section that I'm not aware of. Can you please share the logs where it doesn't works?

**Comment by konsumer** [comment]
Here is the log, where it opens, and when I close it, it holds the console for like 5 seconds, then exits (so it's probly fine)

```
[0] webui_new_window()...
[0] _webui_init()...
[0] _webui_get_current_path()...
[0] _webui_ptr_add(0x0x7fa4c2809800)... Allocate 2048 bytes
[0] _webui_ptr_add(0x0x600002858080)... Allocate 128 bytes
[0] _webui_get_new_window_number()...
[0] _webui_ptr_add(0x0x7fa4c280a000)... Allocate 2048 bytes
[0] _webui_ptr_add(0x0x7fa4c280a800)... Allocate 2048 bytes
[0] _webui_ptr_add(0x0x7fa4c280b000)... Allocate 2048 bytes
[0] webui_new_window() -> New window @ 0x0x600002858080
[1] webui_show()...
[0] _webui_ptr_add(0x0x6000006591a0)... Allocate 32 bytes
[1] webui_show()... -> Static HTML Script:
- - -[HTML]- - - - - - - - - -
<html>Hello</html>
- - - - - - - - - - - - - - - -
[1] _webui_show_window(html, [0])...
[1] _webui_wait_process()...
[1] webui_is_shown()...
[0] _webui_get_free_port()...
[0] _webui_port_is_used([40720])...
[0] _webui_socket_test_listen_mg([40720])...
[0] _webui_ptr_add(0x0x7fa4c3008200)... Allocate 2048 bytes
[0] _webui_browser_start([http://localhost:40720], [0])...
[0] _webui_browser_start_chrome([http://localhost:40720])...
[0] _webui_browser_exist([1])...
[0] _webui_cmd_sync()...
[0] _webui_cmd_sync() -> Running [open -R -a "Google Chrome" >>/dev/null 2>>/dev/null ]
[0] _webui_browser_create_profile_folder(1)...
[0] _webui_browser_get_temp_path([1])...
[1] _webui_run_browser()...
[0] _webui_ptr_add(0x0x60000065c000)... Allocate 32 bytes
[0] _webui_ptr_add(0x0x7fa4c3804080)... Allocate 512 bytes
[L] webui_wait()...
[L] webui_wait() -> Using timeout 8 second
[0] _webui_wait_for_startup()...
[1] [Thread] webui_server_start(http://localhost:40720)...
[1] _webui_run_browser_detect_proc_task()...
[0] _webui_cmd_sync()...
[0] _webui_cmd_sync() -> Running [/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --user-data-dir="/Users/konsumer/.WebUI/WebUIChromeProfile" --no-first-run --disable-gpu --disable-software-rasterizer --no-proxy-server --safe-mode --disable-extensions --disable-background-mode --disable-plugins --disable-plugins-discovery --disable-translate --bwsi --disable-sync --incognito --app=http://localhost:40720 >>/dev/null 2>>/dev/null ]
[1] [Thread] webui_server_start(http://localhost:40720)... Listening Success -> Loop (8 seconds timeout)...
[0] _webui_wait_for_startup() -> Finish.
[L] webui_wait() -> Wait for connected UI...
[1] _webui_server_event_handler()... HTML Main
[1] _webui_generate_js_bridge()...
[0] _webui_ptr_add(0x0x60000285c000)... Allocate 128 bytes
[0] _webui_ptr_add(0x0x7fa4c4008a00)... Allocate 8192 bytes
[0] _webui_ptr_add(0x0x7fa4c400aa00)... Allocate 8192 bytes
[0] _webui_free_mem(0x0x7fa4c4008a00)... Free 8192 bytes
[0] _webui_free_mem(0x0x7fa4c400aa00)... Free 8192 bytes
[1] _webui_server_event_handler()... HTML Upgrade to WebSocket
[1] _webui_server_event_handler()... WebSocket Connected
[1] _webui_generate_internal_id([])...
[0] _webui_ptr_add(0x0x600000450000)... Allocate 8 bytes
[1] _webui_window_event([1/], [])...
[0] _webui_ptr_add(0x0x60000135c000)... Allocate 64 bytes
[1] [Thread] _webui_cb()...
[1] [Thread] _webui_cb()... Stop.
[0] _webui_free_mem(0x0x600000450000)... Free 8 bytes
[0] _webui_free_mem(0x0x60000135c000)... Free 64 bytes
[1] _webui_server_event_handler()... WebSocket Closed
[1] _webui_generate_internal_id([])...
[0] _webui_ptr_add(0x0x600000454000)... Allocate 8 bytes
[1] _webui_window_event([1/], [])...
[0] _webui_ptr_add(0x0x60000135c000)... Allocate 64 bytes
[1] [Thread] _webui_cb()...
[1] [Thread] _webui_cb()... Stop.
[0] _webui_free_mem(0x0x600000454000)... Free 8 bytes
[0] _webui_free_mem(0x0x60000135c000)... Free 64 bytes
[1] [Thread] webui_server_start()... Server Stop.
[0] _webui_free_port([40720])...
[L] webui_wait() -> Wait finished.
[0] _webui_clean()...
[0] _webui_free_all_mem()...
[0] _webui_free_all_mem()... Free 2048 bytes @ 0x0x7fa4c2809800
[0] _webui_free_all_mem()... Free 128 bytes @ 0x0x600002858080
[0] _webui_free_all_mem()... Free 2048 bytes @ 0x0x7fa4c280a000
[0] _webui_free_all_mem()... Free 2048 bytes @ 0x0x7fa4c280a800
[0] _webui_free_all_mem()... Free 2048 bytes @ 0x0x7fa4c280b000
[0] _webui_free_all_mem()... Free 32 bytes @ 0x0x6000006591a0
[0] _webui_free_all_mem()... Free 2048 bytes @ 0x0x7fa4c3008200
[0] _webui_free_all_mem()... Free 32 bytes @ 0x0x60000065c000
[0] _webui_free_all_mem()... Free 512 bytes @ 0x0x7fa4c3804080
[0] _webui_free_all_mem()... Free 128 bytes @ 0x0x60000285c000
```
After enabling logging, I can't seem to get it to open, then close right away.

Here is the log when it prematurely exits. It displays the content, it just returns the console right away. If I right-click and reload web content, it errors:
```
[0] webui_new_window()...
[0] _webui_init()...
[0] _webui_get_current_path()...
[0] _webui_ptr_add(0x0x7fafdf809800)... Allocate 2048 bytes
[0] _webui_ptr_add(0x0x600001570080)... Allocate 128 bytes
[0] _webui_get_new_window_number()...
[0] _webui_ptr_add(0x0x7fafdf80a000)... Allocate 2048 bytes
[0] _webui_ptr_add(0x0x7fafdf80a800)... Allocate 2048 bytes
[0] _webui_ptr_add(0x0x7fafdf80b000)... Allocate 2048 bytes
[0] webui_new_window() -> New window @ 0x0x600001570080
[1] webui_show()...
[0] _webui_ptr_add(0x0x600003b711a0)... Allocate 32 bytes
[1] webui_show()... -> Static HTML Script:
- - -[HTML]- - - - - - - - - -
<html>Hello</html>
- - - - - - - - - - - - - - - -
[1] _webui_show_window(html, [0])...
[1] _webui_wait_process()...
[1] webui_is_shown()...
[0] _webui_get_free_port()...
[0] _webui_port_is_used([44663])...
[0] _webui_socket_test_listen_mg([44663])...
[0] _webui_ptr_add(0x0x7fafdf80b800)... Allocate 2048 bytes
[0] _webui_browser_start([http://localhost:44663], [0])...
[0] _webui_browser_start_chrome([http://localhost:44663])...
[0] _webui_browser_exist([1])...
[0] _webui_cmd_sync()...
[0] _webui_cmd_sync() -> Running [open -R -a "Google Chrome" >>/dev/null 2>>/dev/null ]
[0] _webui_browser_create_profile_folder(1)...
[0] _webui_browser_get_temp_path([1])...
[1] _webui_run_browser()...
[0] _webui_ptr_add(0x0x600003b74000)... Allocate 32 bytes
[0] _webui_ptr_add(0x0x7fafe0004080)... Allocate 512 bytes
[L] webui_wait()...
[L] webui_wait() -> Using timeout 8 second
[0] _webui_wait_for_startup()...
[1] _webui_run_browser_detect_proc_task()...
[0] _webui_cmd_sync()...
[0] _webui_cmd_sync() -> Running [/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --user-data-dir="/Users/konsumer/.WebUI/WebUIChromeProfile" --no-first-run --disable-gpu --disable-software-rasterizer --no-proxy-server --safe-mode --disable-extensions --disable-background-mode --disable-plugins --disable-plugins-discovery --disable-translate --bwsi --disable-sync --incognito --app=http://localhost:44663 >>/dev/null 2>>/dev/null ]
[1] [Thread] webui_server_start(http://localhost:44663)...
[1] [Thread] webui_server_start(http://localhost:44663)... Listening Success -> Loop (8 seconds timeout)...
[0] _webui_wait_for_startup() -> Finish.
[L] webui_wait() -> Wait for connected UI...
[1] _webui_run_browser_detect_proc_task() -> Process closed.
[0] _webui_free_mem(0x0x7fafe0004080)... Free 512 bytes
[1] _webui_server_event_handler()... HTML Main
[1] _webui_generate_js_bridge()...
[0] _webui_ptr_add(0x0x600001574000)... Allocate 128 bytes
[0] _webui_ptr_add(0x0x7fafe0808a00)... Allocate 8192 bytes
[0] _webui_ptr_add(0x0x7fafe080aa00)... Allocate 8192 bytes
[0] _webui_free_mem(0x0x7fafe0808a00)... Free 8192 bytes
[0] _webui_free_mem(0x0x7fafe080aa00)... Free 8192 bytes
[1] _webui_server_event_handler()... HTML Upgrade to WebSocket
[1] _webui_server_event_handler()... WebSocket Connected
[1] _webui_generate_internal_id([])...
[0] _webui_ptr_add(0x0x600003978000)... Allocate 8 bytes
[1] _webui_window_event([1/], [])...
[0] _webui_ptr_add(0x0x600002e74000)... Allocate 64 bytes
[1] [Thread] _webui_cb()...
[1] [Thread] _webui_cb()... Stop.
[0] _webui_free_mem(0x0x600003978000)... Free 8 bytes
[0] _webui_free_mem(0x0x600002e74000)... Free 64 bytes
[1] [Thread] webui_server_start()... Server Stop.
[0] _webui_free_port([44663])...
[L] webui_wait() -> Wait finished.
[0] _webui_clean()...
[0] _webui_free_all_mem()...
[0] _webui_free_all_mem()... Free 2048 bytes @ 0x0x7fafdf809800
[0] _webui_free_all_mem()... Free 128 bytes @ 0x0x600001570080
[0] _webui_free_all_mem()... Free 2048 bytes @ 0x0x7fafdf80a000
[0] _webui_free_all_mem()... Free 2048 bytes @ 0x0x7fafdf80a800
[0] _webui_free_all_mem()... Free 2048 bytes @ 0x0x7fafdf80b000
[0] _webui_free_all_mem()... Free 32 bytes @ 0x0x600003b711a0
[0] _webui_free_all_mem()... Free 2048 bytes @ 0x0x7fafdf80b800
[0] _webui_free_all_mem()... Free 32 bytes @ 0x0x600003b74000
[0] _webui_free_all_mem()... Free 128 bytes @ 0x0x600001574000
```

**Comment by konsumer** [comment]
In all cases, it opens a second chrome icon, so I can get it back to correct stay-open, then clsoe behavior, if I close that in task bar.

This is the webui chrome:
<img width="195" alt="Screenshot 2023-03-31 at 10 23 15 PM" src="https://user-images.githubusercontent.com/83857/229267214-12f565c2-6d8c-4fca-bb59-23747017786c.png">

It stays in taskbar after I exit, and if I run minimal again, it will prematurely exit. If I close it, it will work ok.

**Comment by hassandraga** [comment]
About the 5 seconds, I add it recently to wait if the window reconnect, in case of a refresh, href click etc... I can decrease it but we should be careful.

About the window stay open, this also is the normal behavior if you build it in debug mode. This keep the window open so we can see the logs in the DevTools for the development purpose. But it should definitely close if you build it in release mode.

About the program exits soon, I see `webui_server_start()... Server Stop` before Websocket disconnect, which is mean there is an issue the the server thread. I will look into it soon.

Thank you for sharing this with us. I appreciate it.

**Comment by konsumer** [comment]
> About the window stay open, this also is the normal behavior if you build it in debug mode. This keep the window open so we can see the logs in the DevTools for the development purpose. But it should definitely close if you build it in release mode.

I built in release-mode, and got the same results. It seems like it's that mac thing where closing the window does not mean "exit the application" like in other OS's. I tried finding a flag to change this behavior (`--disable-zero-browsers-open-for-tests` seemed like it might work) but nothing seems to change this.

Since it uses it's own profile, I think it's a distinct process, so maybe webui could just kill it, on exit? I tried running the chrome-command with `&` at end (which outputs the pid and backrounds) and when I `kill -9` it, the window closes correctly and the icon goes away, without killing other chrome instances.

**Comment by hassandraga** [comment]
You can now test the [C examples](https://github.com/alifcommunity/webui/tree/main/examples/C/call_c_from_js/macOS/Clang) in macOS.

```
cd examples/C/call_c_from_js/macOS/Clang
make
```

I can say the taskbar icon issue is solved by using `open -W`. But if you didn't quit Chrome, the second execution will show the home page. I will look into Chrome commands how we can solve this.

**Comment by hassandraga** [comment]
I tried `--exit --disable-background-mode` but it does not work.

Also, I tried `/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --user-data-dir="/Users/alif/.WebUI/WebUIChromeProfile" --no-first-run --disable-gpu --disable-software-rasterizer --no-proxy-server --safe-mode --disable-extensions --disable-background-mode --disable-plugins --disable-plugins-discovery --disable-translate --bwsi --disable-sync --disable-sync-preferences --app=https://google.com & chrome_pid=$!; wait $chrome_pid; pkill -P $chrome_pid`.

How do I force Google Chrome to close when no window is left?
